### PR TITLE
fix: IPRPC pool rewards distribution

### PR DESCRIPTION
**Pull request by oren-lava**
# Description

Closes: #XXXX

Found a bug in the IPRPC pool distribution mechanism that ignored some of the leftover coins due to inappropriate error handling.

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] read the [contribution guide](https://github.com/lavanet/lava/blob/main/CONTRIBUTING.md)
* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title, you can find examples of the prefixes below:
    
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the `main` branch
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration tests
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

File: x/rewards/keeper/iprpc.go [hunk 1]
```diff
@@ -148,6 +147,26 @@ func (k Keeper) distributeIprpcRewards(ctx sdk.Context, iprpcReward types.IprpcR
 			continue
 		}
 
+		// remove providers that are registered in the spec fund but unstaked (can't reward unstaked providers)
+		epoch, _, err := k.epochstorage.GetEpochStartForBlock(ctx, uint64(ctx.BlockHeight()))
```
Comment (Yaroms) [inline x/rewards/keeper/iprpc.go:151]: you dont need this, use the current stake storage, not for a specific epoch

Comment (Yaroms) [reply x/rewards/keeper/iprpc.go:151]: also i would advise to do this check when building the `specCuMap` map

Comment (oren-lava) [reply x/rewards/keeper/iprpc.go]: done. see https://github.com/lavanet/lava/pull/1678/commits/0d7f04963d0c02fbc37cc820106d8b2b1afa509b

**Review by Yaroms — DISMISSED** [review]

File: x/dualstaking/keeper/delegator_reward.go [hunk 2]
```diff
@@ -158,39 +159,46 @@ func (k Keeper) ClaimRewards(ctx sdk.Context, delegator string, provider string)
...
 		contributorReward = contributorReward.QuoInt(contributorsNum).MulInt(contributorsNum)
-		claimableRewards = totalReward.Sub(contributorReward...)
 		if !calcOnlyContributor {
 			err = k.PayContributors(ctx, senderModule, contributorAddresses, contributorReward, chainID)
 			if err != nil {
-				return zeroCoins, zeroCoins, err
+				return zeroCoins, claimableRewards, err
 			}
 		}
+		claimableRewards = claimableRewards.Sub(contributorReward...)
```
Comment (omerlavanet) [inline x/dualstaking/keeper/delegator_reward.go:201]: should only be subtracted if we send the rewards inside !calcOnly
we are missing subtracting the rewards on k.rewardDelegator line:216

File: x/rewards/keeper/iprpc.go [hunk 3]
```diff
@@ -178,9 +177,11 @@ func (k Keeper) distributeIprpcRewards(ctx sdk.Context, iprpcReward types.IprpcR
 			UsedReward = UsedRewardTemp
 
 			// reward the provider
-			_, _, err := k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerCU.Provider, specFund.Spec, providerIprpcReward, string(types.IprpcPoolName), false, false, false)
+			_, claimableRewards, err := k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerCU.Provider, specFund.Spec, providerIprpcReward, string(types.IprpcPoolName), false, false, false)
 			if err != nil {
+				// failed sending the rewards, add the claimable rewards to the leftovers that will be transferred to the community pool
 				utils.LavaFormatError("failed to send iprpc rewards to provider", err, utils.LogAttr("provider", providerCU))
+				leftovers = leftovers.Add(claimableRewards...)
```
Comment (omerlavanet) [inline x/rewards/keeper/iprpc.go:184]: let's add a unitest for this

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/dualstaking/keeper/delegator_reward.go (hunk 2 above)
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go:203]: I don't think that's true. In the case of calcOnly, you still want to subtract to get an estimation of the leftovers. By keeping the substraction only inside !calcOnly blocks, the provider monthly payout query won't return a correct result.

Fixed some of the comments to make it clearer here: https://github.com/lavanet/lava/commit/f5ef5ac51a333272552710aba28b8adcbf8874e6

File: x/rewards/keeper/iprpc.go (hunk 3 above)
Comment (omerlavanet) [reply x/rewards/keeper/iprpc.go:184]: if this is unreachable code let's panic instead

Comment (oren-lava) [reply x/rewards/keeper/iprpc.go]: done. see https://github.com/lavanet/lava/commit/0477945caf6cb9cfea6b0c61d8f6caf67141083a

**Comment by oren-lava** [comment]
note that I added a small fix to the `UnstakeEntryForce` function here: https://github.com/lavanet/lava/commit/ea74facb581a258b52164fe8eb6eff4304a96bcf

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**